modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridDhtTxFinishFuture.java

```diff
@@ -485,6 +486,12 @@ private boolean finish(boolean commit,
             req.writeVersion(tx.writeVersion() != null ? tx.writeVersion() : tx.xidVersion());
 
             try {
+                if (isNull(cctx.discovery().node(n.id()))){
```

Comment:
Seems that it's not valid. If we will use some other implementation of CommunicationSPI instead of default one we might override node() method of IgniteSpiContext that won't use `ctx.discovery().node(nodeId)`.